File: part2_section3/cpp_scripts/memory_management.cpp
#include <iostream>
#include <vector>
#include <memory>

std::vector<int>* createLargeArray() {
    std::cout << "Creating large array" << std::endl;
    // Allocate a vector with 1 million integers (~4MB)
    std::vector<int>* data = new std::vector<int>();
    data->reserve(1000000);
    std::cout << "Array created with capacity: " << data->capacity() << std::endl;
    return data;
}

void processData(const std::vector<int>* data) {
    std::cout << "Processing data (pointer) of capacity: " << data->capacity() << std::endl;
    // In C++, we're working with a pointer which must be managed
}

int main() {
    std::cout << "C++ memory management demo" << std::endl;
    
    {
        // Create data - manually managed
        std::vector<int>* largeData = createLargeArray();
        
        // Pass the pointer
        processData(largeData);
        
        std::cout << "End of inner scope - must manually delete largeData" << std::endl;
        
        // Manual memory management requires explicit deallocation
        delete largeData;
    } // largeData pointer goes out of scope, but memory would leak without delete
    
    std::cout << "Memory has been manually freed" << std::endl;
    
    return 0;
}

File: part2_section3/java_scripts/MemoryManagement.java
import java.util.ArrayList;

public class MemoryManagement {
    
    private static ArrayList<Integer> createLargeArray() {
        System.out.println("Creating large array");
        // Allocate array list with initial capacity for 1 million integers
        ArrayList<Integer> data = new ArrayList<>(1_000_000);
        System.out.println("Array created with capacity: " + data.size());
        return data;
    }
    
    private static void processData(ArrayList<Integer> data) {
        System.out.println("Processing data (reference) of capacity: " + data.size());
        // In Java, references are passed but garbage collection handles memory
    }
    
    public static void main(String[] args) {
        System.out.println("Java memory management demo");
        
        {
            // Create data - managed by garbage collector
            ArrayList<Integer> largeData = createLargeArray();
            
            // Pass the reference
            processData(largeData);
            
            System.out.println("End of inner block - reference will go out of scope");
        } // largeData reference goes out of scope, but memory not immediately freed
        
        System.out.println("Memory will be freed by garbage collector when needed");
        
        // Trigger garbage collection (just for demonstration)
        System.gc();
    }
}